### PR TITLE
RFC/WIP: Experimental backward time integration

### PR DESCRIPTION
**Pull request by mauro3**

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -41,11 +41,22 @@ type IVP{T,Y,F,G,J} <: AbstractIVP{T,Y}
     F!  ::F
     G!  ::G
     J!  ::J
+    reversed::Bool
```
Comment (mauro3) [inline src/types.jl:44]: I think it's kinda awkward to have this here.  We decided that only `t0` but not `tend` belongs in this type.  I think the same can be said for including the time direction.  Move it to `Solver`?

Comment (pwl) [inline src/types.jl:44]: I totally agree, but it would still look awkward there.

File: src/runge-kutta.jl [hunk 2]
```diff
@@ -46,8 +46,19 @@ name(stepper::RKStepper) = stepper.tableau.name
...
-solve{T,S<:RKStepper}(ode::ExplicitODE{T}, stepper::Type{S}; options...) =
-    Solver(ode,stepper{T}(;options...))
+function solve{T,S<:RKStepper}(ode::ExplicitODE{T},
+                               ::Type{S};
+                               tspan = T[Inf],
+                               tstop = tspan[end],
+                               options...)
+    t0 = ode.t0
+    if tstop < t0
+        tspan = 2t0.-tspan
```
Comment (mauro3) [inline src/runge-kutta.jl:56]: Why the `2`?

Comment (pwl) [inline src/runge-kutta.jl:56]: if `tspan = [t0,t0+dt]` (with `dt<0`), then an interval of the same length and starting at `t0`, but reversed, would be `[t0,t0-dt]`.  But `dt=tstop-t0` and the same for `tspan`.

**Comment by mauro3** [comment]
I think if this where to be done, it would need to go into the `Solver` type, have a `tend` field there.  If the problem is reversed then create a new `IVP` type with the reversal transformations applied in the `start` method.  In the `next` method check for reversal and return `-t` if necessary.  Outside of `start`, `done` and `next` nothing should be aware of this trick.  Then add something like the `onestep` etc. bits to the iterator, so solver-implementations would not need to deal with this at all.

Anyway, I still think it is not worth the hassle.  This is essentially weighting up between the [KISS](https://en.wikipedia.org/wiki/KISS_principle) and the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle: the KISS solution adds `tdir` to comparisons in the solvers, the DRY drops that repetition by some complicated refactor and wrapping.  I think the little bit of DRY-ness is not worth the big loss of KISS-ness.  We've spent several man-hours on this already and it is not done yet...

**Comment by pwl** [comment]
You can't have it in `start`, `IVP` and `Solver` are immutable and you have to reverse an ode in the constructor of `Solver` (or in `solve`).  Anyway, I agree that some options should be kept in `Solver` (for example, we have `tstop` in three places right now).  I also commented on it on gitter.

**Comment by mauro3** [comment]
I think you would need to have the `state` carry a `IVP` instance, either the original one or a reversed one.  I think the fact that dense-output needs to be aware of this is not ok.

**Comment by pwl** [comment]
I agree that the level of complexity added is too great at this point.  It was worth a shot anyway.

**Comment by mauro3** [comment]
Cool.  And yes, I think `tstop` would fit well with `Solver`.

**Pull request closed without merging**